### PR TITLE
feat: support new Bootstrapped and Errored watch events

### LINKED ISSUE #175: Watch API improvements
Introduce new types of Watch events:

* Error to pass through errors, we need to use it in the gRPC wrapper, in the etcd state implementation; inmem implementation can use it to report buffer overrun
* Bootstrapped to report when initial contents are sent (only when BootstrapContents is requested)

Figure out how to introduce it without breaking clients... New Watch call? add options to the existing one?

### PR DESCRIPTION
**Pull request by smira**
In the regular COSI, new event types are enabled immediately.

With protobuf client, new types are enabled when using new COSI client, old COSI client against new server won't receive new events, as old implementation panics on nil Resource in the event structure.

Fixes #175

Signed-off-by: Andrey Smirnov <andrey.smirnov@talos-systems.com>

### CODE REVIEW

**Comment by smira** [comment]
See https://github.com/cosi-project/specification/pull/5

**Review by TimJones — APPROVED** [review]

File: pkg/state/impl/inmem/collection.go [hunk 1]
```diff
@@ -455,10 +465,15 @@ func (collection *ResourceCollection) WatchAll(ctx context.Context, ch chan<- st
 			}
 
 			if collection.writePos-pos >= int64(collection.capacity) {
-				// buffer overrun, there's no way to signal error in this case,
-				// so for now just return
 				collection.mu.Unlock()
 
+				select {
+				case ch <- state.Event{
+					Type: state.Errored,
```
Comment (utkuozdemir) [inline pkg/state/impl/inmem/collection.go:472]: why don't we fill the error field when the state is errored?

Comment (smira) [reply pkg/state/impl/inmem/collection.go:472]: :facepalm:  thanks for noticing that, I will write a test for it!

Comment (utkuozdemir) [reply pkg/state/impl/inmem/collection.go:472]: I think the assertions need to be inverted as well :)

**Review by utkuozdemir — APPROVED** [review]
🆒

**Comment by smira** [comment]
I addressed the feedback, and added tests for Watch errors (simulated buffer overrun), and also controller-runtime now fails if any internal watch call fails (also with a unit-test)

**Pull request merged**